File: nvpl_blas/c/chpr.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_chpr
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t incX = 1;
    float alpha = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    nvpl_scomplex_t * A = NULL;
    nvpl_scomplex_t * X = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);

    printf("\nExample: cblas_chpr for the Hermitian rank-1 update\n\n");
    printf("#### args: n=%" PRId64 ", incX=%" PRId64 ", order=%c, uplo=%c\n",
           (int64_t)N, (int64_t)incX, order_to_char(order), uplo_to_char(uplo));
    printf("           alpha=%g\n", alpha);

    // allocate memory
    A = (nvpl_scomplex_t *)malloc(N * N  * sizeof(nvpl_scomplex_t));
    X = (nvpl_scomplex_t *)malloc(len_x * sizeof(nvpl_scomplex_t));

    // fill data
    fill_cmatrix(A, N, N, N, order, UPLO_TO_FILL_MODE(uplo), CblasUnit);
    fill_cvector(X, N, incX);

    // print input data
    print_cmatrix(A, N, N, N, "A", order);
    print_cvector(X, N, incX, "X");

    // call cblas_chpr
    cblas_chpr(order, uplo, N, alpha, X, incX, A);

    // print result
    printf("\nAfter chpr operation:\n");
    print_cmatrix(A, N, N, N, "A", order);

    // release memory
    free(A);
    free(X);
    return 0;
}

File: nvpl_blas/f77/chpr.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     CHPR 
!
!******************************************************************************
      program   CHPR_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, i
      parameter       (n=2, incx=1)
      real            alpha
      parameter       (alpha=1.0)
      integer         xsize
      parameter       (xsize=1+(n-1)*abs(incx))
      complex         x(xsize), a(n,n)

*       External Subroutines
      external        print_cvector, fill_cvector, fill_cmatrix,
     &                print_cmatrix
      character*1     uplo
      parameter       (uplo='U')
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx
      print 101, alpha
      call fill_cvector(x,n,incx)
      call fill_cmatrix(a,n,n,n,uplo,'U')
      call print_cmatrix(a,n,n,n,'A')
      call print_cvector(x,n,incx,'X')
*
*      Call CHPR subroutine
      call CHPR(uplo,n,alpha,x,incx,a)
*
      print*
      print 102
      call print_cmatrix(a,n,n,n,'A')

      stop
 99   format('Example: CHPR for the packed Hermitian rank-1 update')
 100  format('#### args: n=',i1,', incx=',i1)
 101  format(11x,'alpha=',f3.1)
 102  format('The result of CHPR: ')
      end
